File: Session/TSSTThumbnailView.m
//
//  TSSTThumbnailView.m
//  SimpleComic
//
//  Created by Alexander Rauchfuss on 8/22/07.
//  Copyright 2007 Dancing Tortoise Software. All rights reserved.
//

#import "TSSTThumbnailView.h"
#import "TSSTSessionWindowController.h"
#import "TSSTImageUtilities.h"
#import "TSSTImageView.h"
#import "TSSTInfoWindow.h"


@implementation TSSTThumbnailView


- (void)awakeFromNib
{
    [[self window] makeFirstResponder: self];
    [[self window] setAcceptsMouseMovedEvents: YES];
    [thumbnailView setClears: YES];
}



- (id)initWithFrame:(NSRect)frame
{
    self = [super initWithFrame:frame];
    if (self)
    {
        hoverIndex = NSNotFound;
        trackingRects = [NSMutableIndexSet new];
        trackingIndexes = [NSMutableSet new];
        threadIdent = 0;
        thumbLock = [NSLock new];
    }
    return self;
}



- (void) dealloc
{
    [thumbLock release];
    thumbLock = nil;
    [trackingRects release];
    [trackingIndexes release];
    [super dealloc];
}



- (NSRect)rectForIndex:(NSInteger)index
{
    NSRect bounds = [[[self window] screen] visibleFrame];
    float ratio = (NSHeight(bounds)) / NSWidth(bounds);
    NSInteger horCount = ceilf(sqrtf([[pageController content] count] / ratio));
    NSInteger vertCount = ceilf((float)[[pageController content] count] / (float)horCount);
    float side = NSHeight(bounds) / vertCount;
    float horSide = NSWidth(bounds) / horCount;
    NSInteger horGridPos = index % horCount;
    NSInteger vertGridPos = (index / horCount) % vertCount;
    
    NSRect thumbRect;
    if([[[dataSource session] valueForKey: @"pageOrder"] boolValue])
    {
        thumbRect = NSMakeRect(horGridPos * horSide, NSMaxY(bounds) - side - vertGridPos * side, horSide, side);
    }
    else
    {
        thumbRect = NSMakeRect(NSMaxX(bounds) - horSide - horGridPos * horSide, NSMaxY(bounds) - side - vertGridPos * side, horSide, side);
    }
    return thumbRect;
}



- (void)setDataSource:(id)source
{
    dataSource = source;
}



- (id)dataSource
{
    return dataSource;
}



- (void)removeTrackingRects
{
    [thumbnailView setImage: nil];
    hoverIndex = NSNotFound;
    NSInteger tagIndex = [trackingRects firstIndex];
    while (tagIndex != NSNotFound)
    {
        [self removeTrackingRect: tagIndex];
        tagIndex = [trackingRects indexGreaterThanIndex: tagIndex];
    }
    [trackingRects removeAllIndexes];
    [trackingIndexes removeAllObjects];
}



- (void)buildTrackingRects
{
    hoverIndex = NSNotFound;
    [self removeTrackingRects];
	NSInteger counter = 0;
	NSRect trackRect;
	NSNumber * rectIndex;
    NSInteger tagIndex;
	for (; counter < ([[pageController content] count]); ++counter)
	{
		trackRect = NSInsetRect([self rectForIndex: counter], 2, 2);
		rectIndex = @(counter);
		tagIndex = [self addTrackingRect: trackRect 
								   owner: self 
								userData: rectIndex
							assumeInside: NO];
		[trackingRects addIndex: tagIndex];
		[trackingIndexes addObject: rectIndex];
	}
    [self setNeedsDisplay: YES];
}



- (void)drawRect:(NSRect)rect
{
    NSImage * thumbnail;
    NSRect drawRect;
    NSInteger counter = 0;
    NSPoint mouse = [NSEvent mouseLocation];
    NSRect point = NSMakeRect(mouse.x, mouse.y, 6.0f, 6.0f);
    NSPoint mousePoint = [[self window] convertRectFromScreen: point].origin;
	mousePoint = [self convertPoint: mousePoint fromView: nil];
    while (counter < limit)
    {
        thumbnail = [dataSource imageForPageAtIndex: counter];
        drawRect = [self rectForIndex: counter];
        drawRect = rectWithSizeCenteredInRect([thumbnail size], NSInsetRect(drawRect, 2, 2));
        [thumbnail drawInRect: drawRect fromRect: NSZeroRect operation: NSCompositeSourceOver fraction: 1.0];
		if(NSMouseInRect(mousePoint, drawRect, NO))
		{
			hoverIndex = counter;
            [self zoomThumbnailAtIndex: hoverIndex];
		}
        ++counter;
    }
}



- (void)mouseDown:(NSEvent *)event
{
    if(hoverIndex < [[pageController content] count] && hoverIndex >= 0)
    {
        [pageController setSelectionIndex: hoverIndex];
    }
    [[self window] orderOut: self];
}



- (void)keyDown:(NSEvent *)event
{
    NSNumber * charNumber = @([[event charactersIgnoringModifiers] characterAtIndex: 0]);
    switch ([charNumber unsignedIntValue])
    {
        case 27:
            [[[self window] windowController] killTopOptionalUIElement];
            break;
        default:
            break;
    }
}



- (void)processThumbs
{
    NSAutoreleasePool * pool = [NSAutoreleasePool new];
    ++threadIdent;
    unsigned localIdent = threadIdent;
    [thumbLock lock];
    NSInteger pageCount = [[pageController content] count];
	NSAutoreleasePool * localPool = [NSAutoreleasePool new];
    limit = 0;
    while(limit < (pageCount) && 
          localIdent == threadIdent && 
          [dataSource respondsToSelector: @selector(imageForPageAtIndex:)])
    {
        [dataSource imageForPageAtIndex: limit];

        
        if(!(limit % 5))
        {
			if([[self window] isVisible])
			{
				[self setNeedsDisplay: YES];
			}
			
			[localPool release];
			localPool = [NSAutoreleasePool new];
        }
        ++limit;
    }
	[localPool release];
    [thumbLock unlock];
    [pool release];
    [self setNeedsDisplay: YES];
}



- (void)mouseEntered:(NSEvent *)theEvent
{
	hoverIndex = [(NSNumber *)[theEvent userData] integerValue];	
    if(limit == [[pageController content] count])
    {
        [NSTimer scheduledTimerWithTimeInterval: 0.05 target: self selector: @selector(dwell:) userInfo: @(hoverIndex) repeats: NO];
    }
}



- (void)mouseExited:(NSEvent *)theEvent
{
    if([(NSNumber *)[theEvent userData] integerValue] == hoverIndex)
    {
        hoverIndex = NSNotFound;
        
		[thumbnailView setImage: nil];
		[[self window] removeChildWindow: [thumbnailView window]];
		[[thumbnailView window] orderOut: self];
    }
}



- (void)dwell:(NSTimer *)timer
{
     if([[timer userInfo] integerValue] == hoverIndex)
     {
         [self zoomThumbnailAtIndex: hoverIndex];
     }
}



- (void)zoomThumbnailAtIndex:(NSInteger)index
{
    NSImage * thumb = [[pageController arrangedObjects][index] valueForKey: @"pageImage"];
	[thumbnailView setImage: thumb];
	[thumbnailView setNeedsDisplay: YES];

    NSSize imageSize = [thumb size];
    thumbnailView.imageName = [[pageController arrangedObjects][index] valueForKey: @"name"];
    NSRect indexRect = [self rectForIndex: index];
    NSRect visibleRect = [[[self window] screen] visibleFrame];
    NSPoint thumbPoint = NSMakePoint(NSMinX(indexRect) + NSWidth(indexRect) / 2,
                                     NSMinY(indexRect) + NSHeight(indexRect) / 2);
    float viewSize = 312;//[thumbnailView frame].size.width;
    float aspect = imageSize.width / imageSize.height;
    
    if(aspect <= 1)
    {
        imageSize = NSMakeSize( aspect * viewSize, viewSize);
    }
    else
    {
        imageSize = NSMakeSize( viewSize, viewSize / aspect);
    }
    
    if(thumbPoint.y + imageSize.height / 2 > NSMaxY(visibleRect))
    {
        thumbPoint.y = NSMaxY(visibleRect) - imageSize.height / 2;
    }
    else if(thumbPoint.y - imageSize.height / 2 < NSMinY(visibleRect))
    {
        thumbPoint.y = NSMinY(visibleRect) + imageSize.height / 2;
    }
    
    if(thumbPoint.x + imageSize.width / 2 > NSMaxX(visibleRect))
    {
        thumbPoint.x = NSMaxX(visibleRect) - imageSize.width / 2;
    }
    else if(thumbPoint.x - imageSize.width / 2 < NSMinX(visibleRect))
    {
        thumbPoint.x = NSMinX(visibleRect) + imageSize.width / 2;
    }
	
    [(TSSTInfoWindow *)[thumbnailView window] setFrame: NSMakeRect(thumbPoint.x - imageSize.width / 2, thumbPoint.y - imageSize.height / 2, imageSize.width, imageSize.height)
											   display: NO
											   animate: NO];
	[[self window] addChildWindow: [thumbnailView window] ordered: NSWindowAbove];
}



@end

File: Classes/Session/TSSTThumbnailView.swift
//
//  TSSTThumbnailView.swift
//  SimpleComic
//
//  Created by C.W. Betts on 10/25/15.
//
//

import Cocoa

class TSSTThumbnailView: NSView {
	@IBOutlet weak var pageController: NSArrayController!
	@IBOutlet weak var thumbnailView: TSSTImageView!
	
	weak var dataSource: TSSTSessionWindowController?
	
	private let trackingRects = NSMutableIndexSet()
	private var trackingIndexes = Set<NSNumber>()
	
	private var hoverIndex: Int? = nil
	private var limit = 0
	
	private var thumbLock = NSLock()
	private var threadIdent: UInt32 = 0;

	override func awakeFromNib() {
		super.awakeFromNib()
		self.window?.makeFirstResponder(self)
		self.window?.acceptsMouseMovedEvents = true
		thumbnailView.clears = true
	}
	
	func rectForIndex(index: Int) -> NSRect {
		let bounds = window!.screen!.visibleFrame
		let ratio = bounds.height / bounds.width
		let horCount = Int(ceil(sqrt(CGFloat(pageController!.content!.count) / ratio)))
		let vertCount = Int(ceil(CGFloat(pageController!.content!.count) / CGFloat(horCount)))
		let side = bounds.height / CGFloat(vertCount)
		let horSide = bounds.width / CGFloat(horCount)
		let horGridPos = index % horCount
		let vertGridPos = (index / horCount) % vertCount
		let thumbRect: NSRect
		if dataSource!.session?.valueForKey("pageOrder")?.boolValue ?? false {
			thumbRect = NSMakeRect(CGFloat(horGridPos) * horSide, NSMaxY(bounds) - side - CGFloat(vertGridPos) * side, horSide, side)
		}
		else {
			thumbRect = NSMakeRect(NSMaxX(bounds) - horSide - CGFloat(horGridPos) * horSide, NSMaxY(bounds) - side - CGFloat(vertGridPos) * side, horSide, side)
		}
		return thumbRect
	}
	
	func removeTrackingRects() {
		thumbnailView.image = nil
		hoverIndex = nil
		var tagIndex: Int = trackingRects.lastIndex
		while tagIndex != NSNotFound {
			self.removeTrackingRect(tagIndex)
			tagIndex = trackingRects.indexLessThanIndex(tagIndex)
		}
		trackingRects.removeAllIndexes()
		trackingIndexes.removeAll()
	}
	
	func buildTrackingRects() {
		hoverIndex = nil
		removeTrackingRects()
		var trackRect: NSRect
		var rectIndex: NSNumber
		for counter in 0 ..< pageController!.content!.count {
			trackRect = rectForIndex(counter).insetBy(dx: 2, dy: 2)
			rectIndex = counter
			let tagIndex = addTrackingRect(trackRect, owner: self, userData: UnsafeMutablePointer(unsafeAddressOf(rectIndex)), assumeInside: false)
			trackingRects.addIndex(tagIndex)
			trackingIndexes.insert(rectIndex)
		}
		needsDisplay = true
	}
	
	func processThumbs() {
		autoreleasepool() {
			++threadIdent
			let localIdent = threadIdent
			thumbLock.lock()
			let pageCount: Int = pageController!.content!.count
			limit = 0
			while limit < (pageCount) && localIdent == threadIdent && dataSource!.respondsToSelector("imageForPageAtIndex:") {
				autoreleasepool() {
					dataSource!.imageForPageAtIndex(limit)
					if (limit % 5) == 0 {
						if window!.visible {
							needsDisplay = true
						}
					}
					++limit
				}
			}
			thumbLock.unlock()
		}
		needsDisplay = true
	}
	
	override func drawRect(rect: NSRect) {
		var counter: Int = 0
		var mousePoint: NSPoint = window!.convertRectFromScreen(NSRect(origin: NSEvent.mouseLocation(), size: .zero)).origin
		mousePoint = convertPoint(mousePoint, fromView: nil)
		while counter < limit {
			let thumbnail = dataSource!.imageForPageAtIndex(counter)
			var drawRect = self.rectForIndex(counter)
			drawRect = rectWithSizeCenteredInRect(thumbnail.size, NSInsetRect(drawRect, 2, 2))
			thumbnail.drawInRect(drawRect, fromRect: NSZeroRect, operation: .CompositeSourceOver, fraction: 1.0)
			if NSMouseInRect(mousePoint, drawRect, false) {
				hoverIndex = counter
				zoomThumbnailAtIndex(hoverIndex!)
			}
			++counter
		}
	}
	
	override func mouseEntered(theEvent: NSEvent) {
		hoverIndex = unsafeBitCast(theEvent.userData, NSNumber.self).integerValue
		if limit == pageController.content?.count {
			NSTimer.scheduledTimerWithTimeInterval(0.05, target: self, selector: "dwell:", userInfo: (hoverIndex! as NSNumber), repeats: false)
		}
	}
	
	override func mouseExited(theEvent: NSEvent) {
		if unsafeBitCast(theEvent.userData, NSNumber.self).integerValue == hoverIndex {
			hoverIndex = nil
			thumbnailView.image = nil
			window!.removeChildWindow(thumbnailView.window!)
			thumbnailView.window!.orderOut(self)
		}
	}
	
	func dwell(timer: NSTimer) {
		if let userInfo = timer.userInfo as? NSNumber, hoverIndex = hoverIndex where userInfo == hoverIndex {
			zoomThumbnailAtIndex(hoverIndex)
		}
	}

	func zoomThumbnailAtIndex(index: Int) {
		guard let arrangedObject = (pageController.arrangedObjects as? NSArray)?[index] as? NSObject, thumb = arrangedObject.valueForKey("pageImage") as? NSImage else {
			assert(false, "could not get image at index \(index)")
			return
		}
		thumbnailView.image = thumb
		thumbnailView.needsDisplay = true
		
		var imageSize = thumb.size
		thumbnailView.imageName = arrangedObject.valueForKey("pageImage") as? String
		let indexRect = rectForIndex(index)
		let visibleRect = window!.screen!.visibleFrame
		var thumbPoint = NSPoint(x: indexRect.minX + indexRect.width / 2, y: indexRect.minY + indexRect.height / 2)
		let viewSize: CGFloat = 312 //[thumbnailView frame].size.width;
		let aspect = imageSize.width / imageSize.height
		
		if aspect <= 1 {
			imageSize = NSSize(width: aspect * viewSize, height: viewSize)
		} else {
			imageSize = NSSize(width: viewSize, height: viewSize / aspect)
		}
		
		if thumbPoint.y + imageSize.height / 2 > visibleRect.maxY {
			thumbPoint.y = visibleRect.maxY - imageSize.height / 2;
		}
		else if thumbPoint.y - imageSize.height / 2 < visibleRect.minY {
			thumbPoint.y = visibleRect.minY + imageSize.height / 2;
		}
		
		if thumbPoint.x + imageSize.width / 2 > visibleRect.maxX {
			thumbPoint.x = visibleRect.maxX - imageSize.width / 2;
		} else if thumbPoint.x - imageSize.width / 2 < visibleRect.minX {
			thumbPoint.x = visibleRect.minX + imageSize.width / 2;
		}
		
		thumbPoint.x -= imageSize.width / 2
		thumbPoint.y -= imageSize.height / 2
		(thumbnailView.window as! TSSTInfoWindow).setFrame(NSRect(origin: thumbPoint, size: imageSize), display: false, animate: false)
		window?.addChildWindow(thumbnailView.window!, ordered: .Above)
	}
	
	override func mouseDown(theEvent: NSEvent) {
		if let hoverIndex = hoverIndex where hoverIndex < pageController.content!.count && hoverIndex >= 0 {
			pageController.setSelectionIndex(hoverIndex)
		}
		window?.orderOut(self)
	}
	
	override func keyDown(theEvent: NSEvent) {
		guard let chars = theEvent.charactersIgnoringModifiers else {
			return
		}
		
		let nsChar = (chars as NSString).characterAtIndex(0)
		if nsChar == 27 {
			(window!.windowController! as! TSSTSessionWindowController).killTopOptionalUIElement()
		}
	}
}
